### PR TITLE
Direct upload step functions skeleton

### PR DESCRIPTION
**Pull request by mbarton**
Adds a skeleton for how AWS step functions could manage video uploads. At the moment it simply triggers on an upload, waits for the video chunks in S3 and then fires off the other lambdas (all of which just do dummy logging). The existing video pipeline continues untouched alongside.

The step functions are deployed by Riff-Raff as a separate cloud-formation stack, alongside the main app. There is also a DEV stack and a script to update it.

The [uploader README](https://github.com/guardian/media-atom-maker/blob/mbarton/step-functions/uploader/README.md) goes into a bit more detail.

### CODE REVIEW

File: project/StateMachines.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import sbt._
+import sbt.Keys._
+
+object StateMachines {
```
Comment (jennysivapalan) [inline project/StateMachines.scala]: Isn't this outputting one `StateMachine` rather multiple `StateMachines`?

File: uploader/src/main/resources/cfn-template.yaml [hunk 2]
```diff
@@ -0,0 +1,139 @@
+AWSTemplateFormatVersion: "2010-09-09"
```
Comment (jennysivapalan) [inline uploader/src/main/resources/cfn-template.yaml]: We should document the cloudformation template for uploader is separate to the rest of the stack.

File: uploader/src/main/resources/cfn-template.yaml [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+Description: "A game-changing pipeline for uploading video to the internet"
```
Comment (jennysivapalan) [inline uploader/src/main/resources/cfn-template.yaml]: nitpick: Can we change the description? game-changing should be related to games, speaking of which my softball team won a lot of games this weekend: https://twitter.com/TheMobSoftball/status/864003010548117506 ⚾

File: uploader/src/main/resources/cfn-template.yaml [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        - PolicyName: StatesExecutionPolicy
+          PolicyDocument:
+            Version: "2012-10-17"
+            Statement:
+              - Effect: Allow
+                Action:
+                  - "lambda:InvokeFunction"
+                Resource: "*"
+
+  {{GetChunkFromS3}}
```
Comment (jennysivapalan) [inline uploader/src/main/resources/cfn-template.yaml]: I'm a bit lost, how are these resources being defined? I see they are in `state-machine.json` but how does that get referred to in here?

File: common/src/main/scala/com/gu/media/aws/UploadAccess.scala [hunk 5]
```diff
@@ -1,21 +1,42 @@
...
+import com.amazonaws.services.stepfunctions.model.ListStateMachinesRequest
 import com.gu.media.Settings
+import scala.collection.JavaConverters._
 
 trait UploadAccess { this: Settings with AwsAccess =>
   val userUploadBucket: String = getMandatoryString("aws.upload.bucket")
   val userUploadFolder: String = getMandatoryString("aws.upload.folder")
   val userUploadRole: String = getMandatoryString("aws.upload.role")
 
+  val pipelineName: String = s"VideoPipeline$stage"
```
Comment (mbarton) [inline common/src/main/scala/com/gu/media/aws/UploadAccess.scala]: I'd like to but unfortunately the cloud formation support for state machines doesn't let you set a name apart from the one directly in the YAML, which must be alphanumeric.

File: uploader/src/main/resources/cfn-template.yaml (hunk 3 above)
Comment (mbarton) [inline uploader/src/main/resources/cfn-template.yaml]: Oh oops, that shouldn't have been left in. I'll remove.

:1st_place_medal: ⚾️ Congrats on the results ⚾️ :1st_place_medal:

File: uploader/README.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+This will compile both the cloud-formation template and the lambda code.
+
+The cloud-formation template will be output at:
+
+```
+uploader/target/scala-2.11/resource_managed/main/media-atom-pipeline.yaml
+```
+
+Use the AWS console to update the DEV stack with the new template. Unfortunately this
+cannot be scripted as federated Janus credentials cannot modify IAM roles 😞.
```
Comment (jennysivapalan) [inline uploader/README.md]: We don't give direct access to everyone, is this going to be an issue for us?

Comment (mbarton) [inline uploader/README.md]: Not sure what you mean? Anyone who has developer credentials for the account will be able to do it from the web console.

File: common/src/main/scala/com/gu/media/aws/AwsCredentials.scala [hunk 7]
```diff
@@ -44,10 +44,7 @@ object AwsCredentials {
   }
 
   private def assumeCrossAccountRole(instance: AWSCredentialsProvider, settings: Settings): AWSCredentialsProvider = {
-    val securityTokens = AWSSecurityTokenServiceClientBuilder
-      .standard()
-      .withCredentials(instance)
-      .build()
+    val securityTokens = new AWSSecurityTokenServiceClient(instance)
```
Comment (akash1810) [inline common/src/main/scala/com/gu/media/aws/AwsCredentials.scala]: does this mean we're always using permanent credentials? even in DEV?

File: uploader/src/main/resources/cfn-template.yaml [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    Description: "The S3 bucket where videos are uploaded to"
+    Type: "String"
+  UploadProgressTable:
+    Description: "The Dynamo table to store progress"
+    Type: "String"
+
+Conditions:
+  CreateProdResources: !Equals [!Ref "Stage", "PROD"]
+  CreateCodeResources: !Equals [!Ref "Stage", "CODE"]
+  CreateDevResources:  !Equals [!Ref "Stage", "DEV"]
```
Comment (mbarton) [inline uploader/src/main/resources/cfn-template.yaml]: All credit to membership for working out how to do this: https://github.com/guardian/support-workers/blob/master/cloud-formation/src/cfn-template.yaml

File: common/src/main/scala/com/gu/media/aws/AwsCredentials.scala (hunk 7 above)
Comment (mbarton) [inline common/src/main/scala/com/gu/media/aws/AwsCredentials.scala]: Nope. The static credentials are only for DEV. CODE and PROD use the normal AWS injected ones

File: uploader/README.md (hunk 6 above)
Comment (jennysivapalan) [inline uploader/README.md]: Resolved by talking! 💬

**Pull request merged**